### PR TITLE
Faster joins: Update room stats and the user directory on workers when finishing join

### PR DESCRIPTION
**Pull request by squahtx**
When finishing a partial state join to a room, we update the current
state of the room without persisting additional events. Workers receive
notice of the current state update over replication, but neglect to wake
the room stats and user directory updaters, which then get incidentally
triggered the next time an event is persisted or an unrelated event
persister sends out a stream position update.

We wake the room stats and user directory updaters at the appropriate
time in this commit.

Part of #12814 and #12815.

Signed-off-by: Sean Quah <seanq@matrix.org>

### CODE REVIEW

File: synapse/replication/tcp/client.py [hunk 1]
```diff
@@ -207,6 +207,11 @@ async def on_rdata(
             # we don't need to optimise this for multiple rows.
             for row in rows:
                 if row.type != EventsStreamEventRow.TypeId:
+                    # The row's data is an `EventsStreamCurrentStateRow`.
+                    # When finishing up a partial state join, the current state of a
+                    # room will change without any new events being persisted, so we
+                    # must poke the replication callbacks ourselves.
+                    self.notifier.notify_replication()
```
Comment (MatMaul) [inline synapse/replication/tcp/client.py:214]: This looks like a trick, shouldn't we do that a bit further down in `elif stream_name == UnPartialStatedRoomStream.NAME:` case ?

Comment (squahtx) [reply synapse/replication/tcp/client.py:215]: The room un-partial stating process looks like:

https://github.com/matrix-org/synapse/blob/cdea7c11d082e73606bea5d0462f7971e90d836c/synapse/handlers/federation.py#L1806-L1820

`update_current_state` sends the `EventsStreamCurrentStateRow` over replication, whereas `clear_partial_state_room` sends the `UnPartialStatedRoomStreamRow`.

My thinking is that I

Comment (squahtx) [reply synapse/replication/tcp/client.py:215]: I've reworded the comment to place less emphasis on faster joins.

**Review by DMRobertson — APPROVED** [review]
I guess the point is that this already works---but the worker responsible for fixing up the directories didn't notice it until now?

**Comment by squahtx** [comment]
> I guess the point is that this already works---but the worker responsible for fixing up the directories didn't notice it until now?

It works if you have more than one event persister, but that is a fluke.
Otherwise, the user directory and room stats aren't updated until we next persist an event, which I assume happens frequently enough that the user impact is not noticeable.

**Pull request merged**